imsim/instcat.py

```diff
@@ -399,67 +400,65 @@ def getHourAngle(self, mjd, ra):
+    """"Read the exposure information from the opsim db file.
 
     The objects are handled by InstCatalog.
     """
-    _req_params = { 'file_name' : str, }
+    _req_params = { 'file_name' : str,
+                    'visit' : int,
+                    'snap' : int}
     _opt_params = {}
     _single_params = []
     _takes_rng = False
 
-    def __init__(self, file_name, logger=None):
+    def __init__(self, file_name, visit, snap, logger=None):
         logger = galsim.config.LoggerWrapper(logger)
         self.file_name = file_name
-        self.meta = {}
-
-        logger.warning('Reading instance catalog %s', self.file_name)
-        with fopen(self.file_name, mode='rt') as _input:
-            for line in _input:
-                if line.startswith('#'):  # comments
-                    continue
-                if line.startswith('object'):
-                    # Assumes objects are all at the end.  Is this necessarily true?
-                    break
-
-                key, value = line.split()
-                logger.debug('meta value: %s = %s',key,value)
-                value = float(value)
-                if int(value) == value:
-                    self.meta[key] = int(value)
-                else:
-                    self.meta[key] = float(value)
-
-        logger.warning("Done reading meta information from instance catalog")
+        self.visit = visit
+        self.meta = {'snap' : snap}
+
+        # Query for the info for the desired visit.
+        logger.warning('Reading info from opsim db file %s for visit %s',
+                       self.file_name, self.visit)
+        with sqlite3.connect(self.file_name) as con:
+            table = 'observations'
+            # Read the column names for the observations table
+            sql = f"select name from pragma_table_info('{table}')"
+            columns = [_[0] for _ in con.execute(sql)]
+            # Fill the self.meta dict
+            sql = f"""select {','.join(columns)} from {table} where
+                      observationId={self.visit}"""
+            for key, value in zip(columns, list(con.execute(sql))[0]):
+                self.meta[key] = value
+        logger.warning('Done reading visit info from opsim db file')
+
+        if snap >= self.meta['numExposures']:
+            raise ValueError('Invalid snap value: %d. For this visit, snap < %d' % (snap, self.meta['numExposures']))
 
         # Add a couple derived quantities to meta values
         # Note a semantic distinction we make here:
-        # "filter" is the number 0,1,2,3,4,5 from the input instance catalog.
-        # "band" is the character u,g,r,i,z,y.
+        # "filter" or "band" is a character u,g,r,i,z,y.
         # "bandpass" will be the real constructed galsim.Bandpass object.
-        self.meta['band'] = 'ugrizy'[self.meta['filter']]
-        self.meta['HA'] = self.getHourAngle(self.meta['mjd'], self.meta['rightascension'])
-        self.meta['airmass'] = self.getAirmass(self.meta['altitude'])
```

Comment:
I'm assuming (but wanted to confirm) that this line is removed because 'airmass' is already in the db?

Reply:
Yes, that's right.